WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -158,33 +224,40 @@ private class WeeklyRoundupDataProvider {
             return
         }
 
-        filterWeeklyRoundupEnabledSites(administeredSites, result: result)
+        filterWeeklyRoundupEnabledSites(administeredSites, in: context, result: result)
     }
 
     /// Filters the sites that have the Weekly Roundup notification setting enabled.
     ///
-    private func filterWeeklyRoundupEnabledSites(_ sites: [Blog], result: @escaping (Result<[Blog], Error>) -> Void) {
+    private func filterWeeklyRoundupEnabledSites(_ sites: [Blog], in context: NSManagedObjectContext, result: @escaping (Result<[Blog], Error>) -> Void) {
         let noteService = NotificationSettingsService(coreDataStack: coreDataStack)
 
         noteService.getAllSettings { settings in
-            let weeklyRoundupEnabledSites = sites.filter { site in
-                guard let siteSettings = settings.first(where: { $0.blog == site }),
-                      let pushNotificationsStream = siteSettings.streams.first(where: { $0.kind == .Device }),
-                      let sitePreferences = pushNotificationsStream.preferences else {
-                    return false
-                }
-
-                return sitePreferences["weekly_roundup"] ?? true
+            context.perform {
+                let weeklyRoundupEnabledSites = Self.weeklyRoundupEnabledSites(settings: settings, sites: sites, in: context)
+                result(.success(weeklyRoundupEnabledSites))
             }
-
-            result(.success(weeklyRoundupEnabledSites))
         } failure: { (error: NSError?) in
             let error = DataRequestError.filterWeeklyRoundupEnabledSitesError(error)
             result(.failure(error))
         }
     }
 
-    private func getAllSites() -> Result<[Blog], Error> {
+    static private func weeklyRoundupEnabledSites(
+        settings: [NotificationSettings],
+        sites: [Blog],
+        in context: NSManagedObjectContext) -> [Blog] {
+        return sites.filter { site in
+            guard let siteSettings = settings.first(where: { $0.blogManagedObjectID == site.objectID }),
```

Comment:
The previous code was:

```swift
guard let siteSettings = settings.first(where: { $0.blog == site })
```

The `settings` is an array of `NotificationSettings` instances. And the type has a `blog` property. Those instances are created in `noteService.getAllSettings` method.

So the problem here is that the comparison `$0.blog == site` is performed on a queue different than `$0.blog` context's queue.

The fix was to compare the object ids `$0.blogManagedObjectID == site.objectID`, which is thread-safe.